src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+    {
+        int maxFileForCompaction = getMaxFileCountForCompaction();
+        List<SSTableStats> mergeableLatestTombstones = getMaxOverlappingSSTables(categorizedSSTables.latestTombstones);
+        // if latest tombstones are overlapping too much, merge them without splitting.
+        if (mergeableLatestTombstones.size() > 1)
+        {
+            logger.debug("{}: merging latest tombstone files", tableName);
+            return Optional.of(new SSTablesForCompaction(toSSTableReader(limit(maxFileForCompaction, mergeableLatestTombstones)), true, false));
+        }
+
+        List<OverlappingSet> mergeableNearExpiryTombstones = getDistinctOverlappingSSTables(categorizedSSTables.nearExpiryTombstones);
+        // If nearExpiry tombstones are overlapping too much or if they are wide, merge & split them.
+        for (OverlappingSet os : mergeableNearExpiryTombstones)
+        {
+            if (os.maxOverlap > 1 || os.timeWindowMs.duration > 2 * getWindowSizeInMs())
+            {
+                logger.debug("{}: merging near expiry tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(limit(maxFileForCompaction, os.sstables)), true, true));
+            }
+        }
+
+        List<SSTableStats> mergeableTombstones = getMaxOverlappingSSTables(categorizedSSTables.expiredTombstones);
+        List<SSTableStats> mergeableData = getMaxOverlappingSSTables(categorizedSSTables.data);
+
+        boolean mergingTombstones = mergeableTombstones.size() > mergeableData.size();
+        List<SSTableStats> mergeableSSTables = mergingTombstones ? mergeableTombstones : mergeableData;
+
+        if (mergeableSSTables.size() > 1)
+        {
+            logger.debug("{}: merging {} files", tableName, mergingTombstones ? "expired tombstone" : "data");
+            return Optional.of(new SSTablesForCompaction(toSSTableReader(limit(maxFileForCompaction, mergeableSSTables)), mergingTombstones, true));
+        }
+
+        return Optional.empty();
+    }
+
+    static double getEstimatedGarbage(TombstoneCounts tombstoneCounts, long totalDataPartitions, long totalDataRows, long totalDataSizeOnDisk)
+    {
+        double avgPartitionSize = totalDataPartitions == 0 ? 0 : (double) totalDataSizeOnDisk / totalDataPartitions;
+        double avgRowSize = totalDataRows == 0 ? 0 : (double) totalDataSizeOnDisk / totalDataRows;
+
+        long partitionTombstones = Math.min(tombstoneCounts.partitionTombstones, totalDataPartitions);
+        long rowTombstones = Math.min(tombstoneCounts.rowTombstones, totalDataRows);
+
+        // we are over eastimating the garbage. so assuming that range tombstones are for the remaining undeleted partitions.
+        long rangeTombstones = Math.min(tombstoneCounts.rangeTombstones / 2, totalDataPartitions - partitionTombstones);
+        return (partitionTombstones * avgPartitionSize) + (rowTombstones * avgRowSize) + (rangeTombstones * avgPartitionSize / 3.0);
```

Comment:
Any particular reasoning for factor of 2 and 3 for range tombstones or just a guesstimate?

Reply:
2 is coming from the fact that each range tombstone results in 2 tombstone markers. 1 for the start and another for the end.
3 is coming from the fact that average length of an interval when start & end are chosen randomly from (a,b) is (b-a) / 3.